checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -142,6 +305,94 @@ private void visitAccess(ExpressionTree indexTree, ExpressionTree arrTree) {
     }
 
     @Override
+    public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+        ExecutableElement elt = TreeUtils.elementFromUse(node);
+        if (!PurityUtils.isSideEffectFree(atypeFactory, elt)) {
+            checkAnnotationsInClass(
+                    elt,
+                    null,
+                    node,
+                    getCurrentPath(),
+                    UpperBoundUtil.SideEffectKind.SIDE_EFFECTING_METHOD_CALL);
+        }
+        return super.visitMethodInvocation(node, p);
+    }
+
+    @Override
+    protected void commonAssignmentCheck(
+            Tree varTree, ExpressionTree valueExp, /*@CompilerMessageKey*/ String errorKey) {
+        super.commonAssignmentCheck(varTree, valueExp, errorKey);
+        if (varTree.getKind() == Kind.VARIABLE) {
+            // Not a reassignment, so nothing to check
+            return;
+        }
+        Element elt = InternalUtils.symbol(varTree);
+
+        if (elt != null
+                && (elt.getKind() == ElementKind.FIELD
+                        || elt.getKind() == ElementKind.PARAMETER
+                        || elt.getKind() == ElementKind.LOCAL_VARIABLE)
+                && !ElementUtils.isEffectivelyFinal(elt)) {
+            TypeKind typeKind = InternalUtils.typeOf(varTree).getKind();
+            Receiver field = FlowExpressions.internalReprOf(atypeFactory, (ExpressionTree) varTree);
+            SideEffectKind kind;
+            if (elt.getKind() == ElementKind.FIELD) {
+                if (typeKind == TypeKind.ARRAY) {
+                    kind = UpperBoundUtil.SideEffectKind.ARRAY_FIELD_REASSIGNMENT;
+                } else if (!typeKind.isPrimitive()) {
+                    kind = UpperBoundUtil.SideEffectKind.NON_ARRAY_FIELD_REASSIGNMENT;
+                } else {
+                    return;
+                }
+            } else if (typeKind == TypeKind.ARRAY) {
+                kind = SideEffectKind.LOCAL_VAR_REASSIGNMENT;
+            } else {
+                return;
+            }
+            checkAnnotationsInClass(elt, field, varTree, getCurrentPath(), kind);
+        }
+    }
+
+    private void checkAnnotationsInClass(
+            Element elt,
+            Receiver canonicalTargetName,
+            Tree tree,
+            TreePath path,
+            SideEffectKind sideEffectKind) {
+        List<? extends Element> enclosedElts =
+                ElementUtils.enclosingClass(elt).getEnclosedElements();
+        List<AnnotatedTypeMirror> enclosedTypes = findEnclosedTypes(enclosedElts);
```

Comment:
It seems extremely costly to recalculate this list for each call.  (Plus calling `UpperBoundUtil.getDependentReceivers(atm.getAnnotations(), path, atypeFactory)` on each type every time.) . Can you calculate the list of receivers once per class?

Reply:
Ping